cmd/rpcdaemon/commands/zkevm_api.go

```diff
@@ -230,6 +230,130 @@ func (api *ZkEvmAPIImpl) VerifiedBatchNumber(ctx context.Context) (hexutil.Uint6
 	return hexutil.Uint64(highestVerifiedBatchNo), nil
 }
 
+type BatchDataSlim struct {
+	Number      uint64 `json:"number"`
+	BatchL2Data []byte `json:"batchL2Data,omitempty"`
+	Empty       bool   `json:"empty"`
+}
+
+// GetBatchDataByNumbers returns the batch data for the given batch numbers
+func (api *ZkEvmAPIImpl) GetBatchDataByNumbers(ctx context.Context, batchNumbers []rpc.BlockNumber) (json.RawMessage, error) {
+	tx, err := api.db.BeginRo(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback()
+
+	hermezDb := hermez_db.NewHermezDbReader(tx)
+
+	bds := make([]*BatchDataSlim, 0, len(batchNumbers))
+
+	for _, batchNumber := range batchNumbers {
+		bd := &BatchDataSlim{
+			Number: uint64(batchNumber.Int64()),
+			Empty:  false,
+		}
+
+		// looks weird but we're using the rpc.BlockNumber type to represent the batch number, LatestBlockNumber represents latest batch
+		if batchNumber == rpc.LatestBlockNumber {
+			highestBlock, err := rawdb.ReadLastBlockSynced(tx)
+			if err != nil {
+				return nil, err
+			}
+			highestBatchNo, err := hermezDb.GetBatchNoByL2Block(highestBlock.NumberU64())
+			if err != nil {
+				return nil, err
+			}
+			batchNumber = rpc.BlockNumber(highestBatchNo)
+		}
+
+		batchNo := uint64(batchNumber.Int64())
+
+		_, found, err := hermezDb.GetLowestBlockInBatch(batchNo)
+		if err != nil {
+			return nil, err
+		}
+		if !found {
+			// not found - set to empty and append
+			bd.Empty = true
+			bds = append(bds, bd)
+			continue
+		}
+
+		// block numbers in batch
+		blocksInBatch, err := hermezDb.GetL2BlockNosByBatch(batchNo)
+		if err != nil {
+			return nil, err
+		}
+
+		// collect blocks in batch
+		var batchBlocks []*eritypes.Block
```

Comment:
this seems a repeated block of code, can it brought into its own function?

Reply:
todo'd this one :)